pinot-integration-tests/src/test/java/org/apache/pinot/compat/tests/SegmentOp.java

```diff
@@ -82,14 +107,221 @@ public void setTableConfigFileName(String tableConfigFileName) {
     return true;
   }
+
+  /**
+   * Deletes the segment for the given segment name and table name.
+   * @return true if delete successful, else false.
+   */
+  private boolean deleteSegment() {
+    try {
+      TableConfig tableConfig = JsonUtils.fileToObject(new File(_tableConfigFileName), TableConfig.class);
+      _tableName = tableConfig.getTableName();
+
+      ControllerTest.sendDeleteRequest(ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL)
+          .forSegmentDelete(_tableName, _segmentName));
+      return verifySegmentDeleted();
+    } catch (Exception e) {
+      LOGGER.error("Request to delete the segment {} for the table {} failed.", _segmentName, _tableName, e);
+      return false;
+    }
+  }
+
+  /**
+   * Verify given table name and segment name deleted from the controller.
+   * @return true if no segment found, else false.
+   * @throws IOException
+   * @throws InterruptedException
+   */
+  private boolean verifySegmentDeleted()
+      throws IOException, InterruptedException {
+    long startTime = System.currentTimeMillis();
+    while (getCountForSegmentName() > 0) {
+      if ((System.currentTimeMillis() - startTime) > DEFAULT_MAX_SLEEP_TIME_MS) {
+        LOGGER.error("Delete segment verification failed, count is greater than zero after max wait time {} ms.",
+            DEFAULT_MAX_SLEEP_TIME_MS);
+        return false;
+      }
+      LOGGER.warn("Delete segment verification count greater than zero, will retry after {} ms.",
+          DEFAULT_SLEEP_INTERVAL_MS);
+      Thread.sleep(DEFAULT_SLEEP_INTERVAL_MS);
+    }
+
+    LOGGER.info("Successfully delete the segment {} for the table {}.", _segmentName, _tableName);
+    return true;
+  }
+
+  /**
+   * Retrieve external view for the given table name.
+   * @return TableViews.TableView of OFFLINE and REALTIME segments.
+   */
+  private TableViews.TableView getExternalViewForTable()
+      throws IOException {
+    return JsonUtils.stringToObject(ControllerTest.sendGetRequest(
+        ControllerRequestURLBuilder.baseUrl(ClusterDescriptor.CONTROLLER_URL).forTableExternalView(_tableName)),
+        TableViews.TableView.class);
+  }
+
+  /**
+   * Retrieve the number of segments for OFFLINE which are in state matching the parameter.
+   * @param state of the segment to be verified in the controller.
+   * @return -1 in case of ERROR, 0 if in OFFLINE state else return count of state matching parameter.
+   */
+  private long getSegmentCountInState(String state)
+      throws IOException {
+    final Set<String> segmentState =
+        getExternalViewForTable().offline != null ? getExternalViewForTable().offline.entrySet().stream()
+            .filter(k -> k.getKey().equals(_segmentName)).flatMap(x -> x.getValue().values().stream())
+            .collect(Collectors.toSet()) : Collections.emptySet();
+
+    if (segmentState.contains(CommonConstants.Helix.StateModel.SegmentStateModel.ERROR)) {
```

Comment:
Isnt it better to extract the specific segment instance state from the externalview (helix APIs) and explicitly check for specific state we want? Then we can apply this method to any state we like -- CONSUMING or ONLINE or OFFLINE etc.

So,
1. read externalview into json
2. Extract the key `mapFields` (a map)
3. Extract the key `_segmentName` (another map)
4. Check all values to be the same as desired state

This way, we can re-use the method to check for any state we like (which we will in the future).

thanks

Reply:
Thanks for the clarification. Made changes to return -1 in case of `ERROR`, 1 if all matches the desired state and 0 otherwise.